src/nvim/strings.c

```diff
@@ -1201,45 +1217,63 @@ int vim_vsnprintf(char *str, size_t str_m, const char *fmt, va_list ap,
             if (fmt_spec == 'g' || fmt_spec == 'G') {
               // can't use %g directly, cause it prints "1.0" as "1"
               if ((abs_f >= 0.001 && abs_f < 10000000.0) || abs_f == 0.0) {
-                fmt_spec = 'f';
+                fmt_spec = ASCII_ISUPPER(fmt_spec) ? 'F' : 'f';
               } else {
                 fmt_spec = fmt_spec == 'g' ? 'e' : 'E';
               }
               remove_trailing_zeroes = true;
             }
 
-            if (fmt_spec == 'f' && abs_f > 1.0e307) {
-              // avoid a buffer overflow
-              memmove(tmp, "inf", sizeof("inf"));
-              str_arg_l = sizeof("inf") - 1;
+            if ((fmt_spec == 'f' || fmt_spec == 'F') && abs_f > 1.0e307) {
+              STRCPY(tmp, infinity_str(f > 0.0, fmt_spec,
+                                       force_sign, space_for_positive));
+              str_arg_l = STRLEN(tmp);
+              zero_padding = 0;
             } else {
-              format[0] = '%';
-              int l = 1;
-              if (precision_specified) {
-                size_t max_prec = TMP_LEN - 10;
-
-                // make sure we don't get more digits than we have room for
-                if (fmt_spec == 'f' && abs_f > 1.0) {
-                  max_prec -= (size_t)log10(abs_f);
+              if (isnan(f)) {
+                // Not a number: nan or NAN
+                STRCPY(tmp, ASCII_ISUPPER(fmt_spec) ? "NAN" : "nan");
+                str_arg_l = 3;
+                zero_padding = 0;
+              } else if (isinf(f)) {
+                STRCPY(tmp, infinity_str(f > 0.0, fmt_spec,
+                                         force_sign, space_for_positive));
+                str_arg_l = STRLEN(tmp);
+                zero_padding = 0;
+              } else {
+                format[0] = '%';
+                int l = 1;
```

Comment:
All places where `l` is used expect or are fine with `size_t`. So change `int` to `size_t`.